File: app/src/main/java/org/scp/gymlog/ui/training/TrainingRecyclerViewAdapter.java
package org.scp.gymlog.ui.training;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import org.scp.gymlog.R;
import org.scp.gymlog.databinding.ListElementFragmentHistoryBitBinding;
import org.scp.gymlog.databinding.ListElementFragmentHistoryHeadersBinding;
import org.scp.gymlog.databinding.ListElementFragmentHistoryVariationBinding;
import org.scp.gymlog.model.Bit;
import org.scp.gymlog.ui.training.rows.ITrainingRow;
import org.scp.gymlog.ui.training.rows.ITrainingRow.Type;
import org.scp.gymlog.ui.training.rows.TrainingBitRow;
import org.scp.gymlog.ui.training.rows.TrainingVariationRow;
import org.scp.gymlog.util.DateUtils;
import org.scp.gymlog.util.FormatUtils;
import org.scp.gymlog.util.WeightUtils;

import java.math.BigDecimal;
import java.util.function.BiConsumer;
import java.util.stream.Stream;

public class TrainingRecyclerViewAdapter extends RecyclerView.Adapter<TrainingRecyclerViewAdapter.ViewHolder> {

    private final ExerciseRows rows;
    private final boolean internationalSystem;

    private BiConsumer<Bit, Integer> onClickListener;

    public TrainingRecyclerViewAdapter(ExerciseRows exerciseRows, boolean internationalSystem) {
        this.rows = exerciseRows;
        this.internationalSystem = internationalSystem;
    }

    public void setOnClickListener(BiConsumer<Bit, Integer> onClickListener) {
        this.onClickListener = onClickListener;
    }

    @Override
    public int getItemViewType(int position) {
        switch (rows.get(position).getType()) {
            case BIT:       return R.layout.list_element_fragment_history_bit;
            case VARIATION: return R.layout.list_element_fragment_history_variation;
            default:        return R.layout.list_element_fragment_history_headers;
        }
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == R.layout.list_element_fragment_history_bit)
            return new ViewHolder(
                    ListElementFragmentHistoryBitBinding.inflate(
                            LayoutInflater.from(parent.getContext()), parent, false
                    ));

        if (viewType == R.layout.list_element_fragment_history_variation)
            return new ViewHolder(
                    ListElementFragmentHistoryVariationBinding.inflate(
                            LayoutInflater.from(parent.getContext()), parent, false
                    ));

        return new ViewHolder(
                ListElementFragmentHistoryHeadersBinding.inflate(
                        LayoutInflater.from(parent.getContext()), parent, false
                ));
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        ITrainingRow row = rows.get(position);

        if (row.getType() == Type.HEADER) {
            return;
        }

        if (row.getType() == Type.VARIATION) {
            TrainingVariationRow vRow = (TrainingVariationRow) row;
            holder.mNote.setText(vRow.getVariation().getName());
            return;
        }

        TrainingBitRow bRow = holder.bitRow = (TrainingBitRow) row;
        Bit bit = bRow.getBit();

        BigDecimal weight = WeightUtils.getWeightFromTotal(
                bit.getWeight(),
                rows.getExercise().getWeightSpec(),
                rows.getExercise().getBar(),
                internationalSystem
            );

        holder.mWeight.setText(FormatUtils.toString(weight));
        holder.mReps.setText(String.valueOf(bit.getReps()));
        holder.mNote.setText(String.valueOf(bit.getNote()));

        if (bit.isInstant()) {
            holder.mTime.setText(R.string.symbol_empty);
            setAlpha(holder, 0.4f);
        } else {
            holder.mTime.setText(DateUtils.getTime(bit.getTimestamp()));
            setAlpha(holder, 1f);
        }
    }

    private void setAlpha(final ViewHolder holder, final float alpha) {
        Stream.of(holder.mWeight, holder.mReps, holder.mNote)
                .forEach(v -> v.setAlpha(alpha));
    }

    @Override
    public int getItemCount() {
        return rows.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public TrainingBitRow bitRow;
        public final TextView mWeight;
        public final TextView mReps;
        public final TextView mTime;
        public final TextView mNote;

        public ViewHolder(ListElementFragmentHistoryBitBinding binding) {
            super(binding.getRoot());
            mWeight = binding.weight;
            mReps = binding.reps;
            mTime = binding.time;
            mNote = binding.note;

            itemView.setOnClickListener(v -> {
                if (onClickListener != null) {
                    onClickListener.accept(bitRow.getBit(), rows.indexOf(bitRow));
                }
            });
        }

        public ViewHolder(ListElementFragmentHistoryVariationBinding binding) {
            super(binding.getRoot());
            mWeight = mReps = mTime = null;
            mNote = binding.variationName;
        }

        public ViewHolder(ListElementFragmentHistoryHeadersBinding binding) {
            super(binding.getRoot());
            mWeight = mReps = mTime = mNote = null;
        }

        @Override
        public String toString() {
            return super.toString() + " '" + mWeight.getText() + "'";
        }
    }
}

File: app/src/main/java/org/scp/gymlog/ui/training/TrainingRecyclerViewAdapter.kt
package org.scp.gymlog.ui.training

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import org.scp.gymlog.R
import org.scp.gymlog.databinding.ListitemHistoryBitBinding
import org.scp.gymlog.databinding.ListitemHistoryHeadersBinding
import org.scp.gymlog.databinding.ListitemHistoryVariationBinding
import org.scp.gymlog.model.Bit
import org.scp.gymlog.ui.training.rows.ITrainingRow
import org.scp.gymlog.ui.training.rows.TrainingBitRow
import org.scp.gymlog.ui.training.rows.TrainingVariationRow
import org.scp.gymlog.util.DateUtils.getTimeString
import org.scp.gymlog.util.FormatUtils.bigDecimal
import org.scp.gymlog.util.WeightUtils.calculate
import java.util.function.BiConsumer


class TrainingRecyclerViewAdapter(
    private val rows: ExerciseRows,
    private val internationalSystem: Boolean
) : RecyclerView.Adapter<TrainingRecyclerViewAdapter.ViewHolder>() {

    var onClickListener: BiConsumer<Bit, Int>? = null

    override fun getItemViewType(position: Int): Int {
        return when (rows[position].type) {
            ITrainingRow.Type.HEADER,
            ITrainingRow.Type.HEADER_SUPERSET -> R.layout.listitem_history_headers
            ITrainingRow.Type.BIT,
            ITrainingRow.Type.BIT_SUPERSET -> R.layout.listitem_history_bit
            ITrainingRow.Type.VARIATION -> R.layout.listitem_history_variation
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return when(viewType) {
            R.layout.listitem_history_bit -> ViewHolder(
                ListitemHistoryBitBinding.inflate(
                    LayoutInflater.from(parent.context), parent, false
                )
            )
            R.layout.listitem_history_variation -> ViewHolder(
                ListitemHistoryVariationBinding.inflate(
                    LayoutInflater.from(parent.context), parent, false
                )
            )
            else -> ViewHolder(
                ListitemHistoryHeadersBinding.inflate(
                    LayoutInflater.from(parent.context), parent, false
                )
            )
        }
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val row = rows[position]

        when (row.type) {
            ITrainingRow.Type.HEADER -> {
                holder.mNumber!!.visibility = View.GONE
            }
            ITrainingRow.Type.HEADER_SUPERSET -> { }

            ITrainingRow.Type.VARIATION -> {
                val vRow = row as TrainingVariationRow
                holder.mNote!!.text = vRow.variation.name
            }

            ITrainingRow.Type.BIT -> {
                holder.bitRow = row as TrainingBitRow
                val bit = holder.bitRow!!.bit

                val weight = bit.weight.calculate(
                    bit.variation.weightSpec,
                    bit.variation.bar)

                holder.mNumber!!.visibility = View.GONE
                holder.mWeight!!.bigDecimal = weight.getValue(internationalSystem)
                holder.mReps!!.text = bit.reps.toString()
                holder.mNote!!.text = bit.note

                if (bit.instant) {
                    holder.mTime!!.setText(R.string.symbol_empty)
                    setAlpha(holder, 0.4f)
                } else {
                    holder.mTime!!.text = bit.timestamp.getTimeString()
                    setAlpha(holder, 1f)
                }
            }
            ITrainingRow.Type.BIT_SUPERSET -> {
                holder.bitRow = row as TrainingBitRow
                val bit = holder.bitRow!!.bit

                val weight = bit.weight.calculate(
                    bit.variation.weightSpec,
                    bit.variation.bar)

                holder.mWeight!!.bigDecimal = weight.getValue(internationalSystem)
                holder.mReps!!.text = bit.reps.toString()
                holder.mNote!!.text = bit.note

                val index = rows.variations.indexOf(bit.variation) + 1
                holder.mNumber!!.text = index.toString()

                if (bit.instant) {
                    holder.mTime!!.setText(R.string.symbol_empty)
                    setAlpha(holder, 0.4f)
                } else {
                    if (index == 1) {
                        holder.mTime!!.text = bit.timestamp.getTimeString()
                    } else {
                        holder.mTime!!.setText(R.string.symbol_empty)
                    }
                    setAlpha(holder, 1f)
                }
            }
        }
    }

    private fun setAlpha(holder: ViewHolder, alpha: Float) {
        listOf(holder.mWeight, holder.mReps, holder.mNote, holder.mNumber)
            .forEach { it?.alpha = alpha }
    }

    override fun getItemCount(): Int {
        return rows.size
    }

    inner class ViewHolder : RecyclerView.ViewHolder {
        var bitRow: TrainingBitRow? = null
        val mWeight: TextView?
        val mReps: TextView?
        val mTime: TextView?
        val mNote: TextView?
        val mNumber: TextView?

        constructor(binding: ListitemHistoryBitBinding) : super(binding.root) {
            mWeight = binding.weight
            mReps = binding.reps
            mTime = binding.time
            mNote = binding.note
            mNumber = binding.number

            itemView.setOnClickListener {
                onClickListener?.accept(bitRow!!.bit, rows.indexOf(bitRow!!))
            }
        }

        constructor(binding: ListitemHistoryVariationBinding) : super(binding.root) {
            mTime = null
            mReps = null
            mWeight = null
            mNote = binding.variationName
            mNumber = null
        }

        constructor(binding: ListitemHistoryHeadersBinding) : super(binding.root) {
            mNote = null
            mTime = null
            mReps = null
            mWeight = null
            mNumber = binding.hashColumn
        }

        override fun toString(): String {
            return super.toString() + " '${mWeight!!.text}'"
        }
    }
}